### PR TITLE
DR-1192 Define Git Branch Strategy

### PR DESCRIPTION
**Pull request by prjh2-nhs**
## Description

- Documentation for the git branching strategy

## Sensitive Information Declaration

To ensure the utmost confidentiality and protect your and others privacy, we kindly ask you to NOT including [PII (Personal Identifiable Information) / PID (Personal Identifiable Data)](https://digital.nhs.uk/data-and-information/keeping-data-safe-and-benefitting-the-public) or any other sensitive data in this PR (Pull Request) and the codebase changes. We will remove any PR that do contain any sensitive information. We really appreciate your cooperation in this matter.

- [x] I confirm that neither PII/PID nor sensitive data are included in this PR and the codebase changes.

### CODE REVIEW

File: docs/diagrams/Git_branching_strategy.drawio
Comment (Tomdango) [inline docs/diagrams/Git_branching_strategy.drawio:1]: Tip - if you rename the file to `Git_branching_strategy.drawio.png`, GitHub will render it and you won't need the extra image file alongside.

File: docs/user-guides/Git_branching_strategy.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+- We prefer task over 'feature' as task better encapsulates the short-lived nature of transient branches.
+- It is a task to be delivered (on the way to delivering a full feature) rather than the full feature.
+- Consequently temporary branches created from main should follow this naming pattern: `task/JIRA_REF_Min_three_word_desc`
+  - Where the minimum three word description starts with an initial capital
+  - And the max length for a branch name is 60 characters.
+
+- Valid branch names would be
+  - `task/DR_1_Update_s3_terraform`
+  - `task/DR_9_Add_new_data_item`
```
Comment (Tomdango) [inline docs/user-guides/Git_branching_strategy.md:28]: I'd personally rather keep this as the actual Jira format - i.e. `task/DR-9_Add_new_data_item`.
This makes it much easier to pull the Jira reference out of the branch name using a regex, and stops you having to do some odd formatting with it once you pull it out.

Comment (Tomdango) [reply docs/user-guides/Git_branching_strategy.md:28]: As a mildly controversal comment - do we really care about the last bit of the branch name? Branches come and go, commits last forever.

File: docs/user-guides/Git_branching_strategy.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+- Create a new task branch from develop
+- When development is complete, then create a pull request to merge back into develop.
+
+## To test a ticket
+
+- Complete testing of the ticket on the develop branch
+- If there are test failures, should that be:
+  - Fixed within the current ticket (since QA is not met)?
+  - Or should that be a bug ticket?
```
Comment (Tomdango) [inline docs/user-guides/Git_branching_strategy.md:46]: If it's failing integration tests, it should have never made it into develop. I'd be happy with the fix in that case to go along with the original ticket with the fixes.

If all the integration tests passed, but something was wrong with the functionality/implementation during manual testing, that'd be a case of raising a bug and fixing it separately.

File: docs/user-guides/Git_branching_strategy.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+- (To clarify, in the diagram earlier, “record bug” could mean
+  - To comment on ticket and resolve as part of that ticket, by addressing fixes in a follow up PR using same branch?
+  - Or, to create a new ticket to make the fixes on a new branch?)
+
+## Identify release candidates
+
+- The team will identify the tickets and their subsequent commits for to include in the release
+- A confluence document records the tickets included for the release
+- Merge the subsequent commits into main
+- Where relevant, use feature flagging to merge code to main but in the “off” position
```
Comment (Tomdango) [inline docs/user-guides/Git_branching_strategy.md:56]: Is it worth raising a ticket about implementing feature flags and drop it into the backlog to come up when we need it?

**Review by Tomdango — COMMENTED** [review]
Looks good - added a few general comments

File: docs/user-guides/Git_branching_strategy.md (hunk 1 above)
Comment (AITR1) [reply docs/user-guides/Git_branching_strategy.md:28]: I think the dev agreement was that the branching name would all be '_' re DR-1211 as having the mixture adds confusion.

Comment (timrickwood) [reply docs/user-guides/Git_branching_strategy.md:28]: I have coded around the underscored jira ref in the branch name - I don't think that is much of an issue given team's preference for consistency around use of underscores

File: docs/user-guides/Git_branching_strategy.md (hunk 3 above)
Comment (timrickwood) [reply docs/user-guides/Git_branching_strategy.md:58]: Under To test a ticket could we maybe be clear about what testing is done BEFORE we merge a task branch to develop and after. The wording 'Complete testing...' and the diagram suggest testing has already happened against the branch ? I think we should be clear about that as we want a high degree of confidence in the task code BEFORE we merge it to develop.

File: docs/user-guides/Git_branching_strategy.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    - (so there are failing integration tests),
+    - then resolve the issues as part of the original ticket.
+  - after merge to develop
+    - (so all the integration tests passed),
+    - but manual testing identifies issues with the functionality/implementation,
+    - then raise a bug to fix this separately.
+
+## Identify release candidates
+
+- The team will identify the tickets and their subsequent commits for to include in the release
```
Comment (timrickwood) [inline docs/user-guides/Git_branching_strategy.md:55]: for to ?

File: docs/user-guides/Git_branching_strategy.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+- A confluence document records the tickets included for the release
```
Comment (timrickwood) [inline docs/user-guides/Git_branching_strategy.md:56]: Is it worth detailing this a bit more - eg add release to jira tickets, use JIRA to produce the release notes that we then hold in confluence - I think the current LD team do something very similar.

**Review by timrickwood — APPROVED** [review]

**Pull request merged**